### ISSUE TITLE
Support for AVMATRIX PVS0403U

### ISSUE THREAD

**Issue by jcoutch** [bitfocus/companion-module-requests#936]
As a user, it would be nice to control the [AVMATRIX PVS0403U](https://www.avmatrix.com/products/pvs0403u-portable-10-1-inch-4-ch-sdihdmi-video-switcher/) from Companion.

Ideally, it would be nice if the majority of the functionality could be controlled, but at a bare minimum, if Program/Preview input switching, as well as the cut/auto button worked, that'd be great.

I was able to get a packet capture from the device, mapped out the messages for various buttons, and created a [datasheet](https://github.com/bitfocus/companion-module-requests/files/9782846/PVS0403U.-.Packet.Layout.pdf).  Hopefully this helps speed up development.

Some of the commands require knowledge of the previous states, which could be ascertained using the "get current state" command, and/or by listening to messages being sent back from the console as it's used.

Also, this may resolve #327 if the packets are reasonably the same.

**Comment by jcoutch** [comment]
Update - AVMATRIX sent me the actual API datasheet...which has a lot more detail than mine does:  [PVS0403U-API V3.1.pdf](https://github.com/bitfocus/companion-module-requests/files/9803703/PVS0403U-API.V3.1.pdf)

**Comment by Jumpcutjeff** [comment]
jcoutch, did you ever get this to work on that AVMatrix switcher? I have the same model and can't figure out that confusing API doc from the company. I emailed them and they never replied. I'd love to be able to integrate this switcher. I was able to integrate a switcher from Feelworld and it works great. Thanks!

**Comment by peternewman** [comment]
> jcoutch, did you ever get this to work on that AVMatrix switcher? I have the same model and can't figure out that confusing API doc from the company.

It doesn't seem too bad @Jumpcutjeff , but I don't have the switcher to test it against.

Have you tried sending some of their example commands?

If you look at their example for "Set PIP1 Source to color bar":
`5A 0E 00 00 00 00 05 24 01 00 00 25 B7 DD`

Which breaks down as:
```
5A - Header
0E 00 - Length (little endian)
00 - Device type
00 - Device ID
00 - Reserved
05 - Frame data length
--- Frame data length start
24 - PIP1 command
01 - Source colour bars
00 - Size 1/2
00 - X=0
25 - Y=25 (hex)=37 decimal
--- Frame data length end
B7 - Checksum (5A+0E+00+00+00+00+05+24+01+00+00+25)
DD - End of frame
```

**Comment by jcoutch** [comment]
@Jumpcutjeff - I was able to get commands to work using a hacked together
C# app, but didn't have time to implement a Companion module.

**Comment by Jumpcutjeff** [comment]
@jcoutch @peternewman Thanks so much for your reply! I tried using their sample code: 5A 0E 00 00 00 00 05 24 01 00 00 25 B7 DD
BUT it didn't work. However, it could be user error. I was using the generic TCP/UDP module for Companion. See pic for how I entered code (not above code). Perhaps I entered it incorrectly? 

<img width="569" alt="UDP-Command" src="https://github.com/bitfocus/companion-module-requests/assets/140195030/d0102397-d9db-4bd2-8412-e54774adbe25">

**Comment by Jumpcutjeff** [comment]
@peternewman I'm absolutely confounded by how to generate the codes from their data sheet. I don't even know what this kind of data is called...hex code? For example, in their API doc, under Transitions->PGM-Source:
Command=0x12
Data=1 (PGM1=1)
Data Length=1

I have no idea of how to use your guide, which I appreciate you posting, to convert those values into a string that resembles the example. Any additional guidance would be appreciated. Thank you!


> Have you tried sending some of their example commands?
> 
> If you look at their example for "Set PIP1 Source to color bar": `5A 0E 00 00 00 00 05 24 01 00 00 25 B7 DD`
> 
> Which breaks down as:
> 
> ```
> 5A - Header
> 0E 00 - Length (little endian)
> 00 - Device type
> 00 - Device ID
> 00 - Reserved
> 05 - Frame data length
> --- Frame data length start
> 24 - PIP1 command
> 01 - Source colour bars
> 00 - Size 1/2
> 00 - X=0
> 25 - Y=25 (hex)=37 decimal
> --- Frame data length end
> B7 - Checksum (5A+0E+00+00+00+00+05+24+01+00+00+25)
> DD - End of frame
> ```

**Comment by peternewman** [comment]
> Thanks so much for your reply! I tried using their sample code: 5A 0E 00 00 00 00 05 24 01 00 00 25 B7 DD BUT it didn't work. However, it could be user error. I was using the generic TCP/UDP module for Companion. See pic for how I entered code (not above code). Perhaps I entered it incorrectly?

I'm a bit confused how you got from my example code to what's in your input box?

Are you sure you're sending to port 19523?

I think you want "none" for command end character.

Looking at https://github.com/bitfocus/companion-module-generic-tcp-udp/issues/3 you just want a % before each hex pair I mentioned, so you actually want to put this in the box:
`%5A%0E%00%00%00%00%05%24%01%00%00%25%B7%DD`

> @peternewman I'm absolutely confounded by how to generate the codes from their data sheet. I don't even know what this kind of data is called...hex code? For example, in their API doc, under Transitions->PGM-Source: Command=0x12 Data=1 (PGM1=1) Data Length=1

Yeah it's hex codes, there guide isn't the clearest, but it's not the worst I've ever seen. As long as you can view PIP1, I'd start with that example as there's a good chance it will actually work if you send it properly. The further you drift from that the greater the chances of it not even being a valid command to send...

No worries, if that string works I can explain more about the mapping stuff.

**Comment by Jumpcutjeff** [comment]
Thanks! Yes, I'm sending to port 19523. I'll try adding those percentage signs and see where we're at.

**Comment by Jumpcutjeff** [comment]
Nothing. I tried enclosing the string in <>, too.

I plugged the switcher into the network on a PC and ran AVMatrix's control software. It did connect. So, it can be controlled via IP. I'm just unsure how to get Companion to connect.

**Comment by jcoutch** [comment]
@Jumpcutjeff - I tried looking for my .NET code, but not sure where it's at.  I did find a sample Node app that I was using during my initial reverse engineering of the protocol, and uploaded it as a Gist:
https://gist.github.com/jcoutch/fc3b37309f1d09f864b839337de90500

And here's a small fiddle I wrote for building the messages:
https://jsfiddle.net/jcoutch/s4gyumn9/26/

Just edit the frame data section, and you should be able to test with different messages.

**Comment by Jumpcutjeff** [comment]
@jcoutch Wow, that looks awesome. Forgive me, but I'am a severe coding dunce. :) Where/how do I run the index.mjs file? I'm on Mac.

**Comment by Jumpcutjeff** [comment]
@jcoutch Ha! This dummy learned a new trick! I installed Visual Studio Code and ran your .mjs file. I copied some of the commands after the [ & it worked!

**Comment by Jumpcutjeff** [comment]
WOW! It's working in Companion. Using the TCP/UDP module, I enter the code: %5a%0c%00%00%00%00%03%11%01%00%7b%dd 
and VOILA! It does an auto transition. This is a game changer. 

Thank you both!

**Comment by peternewman** [comment]
Looking a bit further, my/their PIP example might have been a poor choice as you probably need to turn the PIP on to see the effects.

Looking at @jcoutch it looks like `5a0b0000000002fe0166dd` would get tally feedback, at which point you'd definitely want a module I think...

**Comment by silvansan** [comment]
Yes, please. If this feature is still possible, for people who aren't as coding-savvy as others, this feature on the Bitfocus Companion would *literally* save lives! (Not literally literally lives, but at least some precious time, indeed!)

Please keep me updated if there is a way to more-or-less easily implement this in Companion! I would be grateful, volumes and volumes! 

In the mean time, I'll try to make the TCP/UDP module work :)

**Comment by gianni244** [comment]
have you any news about  the avmatrix connection with companion?

**Comment by silvansan** [comment]
> have you any news about the avmatrix connection with companion?

[WORKS ! bitfocus-companion-hex-commands.txt](https://github.com/user-attachments/files/17269086/WORKS.bitfocus-companion-hex-commands.txt)


I finally took some time to ask Claude.ai, upload the API pdf, and I got it to makes me a list of few **commands in hex** which work.
you have to send the commands to **_tcp-udp: Send HEX encoded Command_**
and just input `5a0c00000000031204007fdd`

Now I wonder if someone knows how to create a module on bitfocus companion, and add all those lines in a module, etc. :)
